### PR TITLE
DOC: Update cilium contributing docs:

### PR DESCRIPTION
**Pull request by eloycoto**
- Added documentation for JustAfterEach, AfterFailed and example flow of
execution.
- Extend documentation related to sshConfig
- Added documenation how to run ginkgo using delve.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

### CODE REVIEW

File: Documentation/contributing.rst [hunk 1]
```diff
@@ -622,6 +622,76 @@ common unit test frameworks.
 A good use case for using ``AfterAll`` method is to remove containers or pods
 that are needed for multiple ``Its`` in the given ``Context`` or ``Describe``.
 
+JustAfterEach
+^^^^^^^^^^^^^
+
+This method will run just after each test and before ``AfterFailed`` and
+``AfterEach``. The main reason of this method is to provide a mechanism to do
+some assertions, for a group of test, after the test finished without
```
Comment (ianvernon) [inline Documentation/contributing.rst]: to perform some assertions for a group of tests.

File: Documentation/contributing.rst [hunk 2]
```diff
@@ -622,6 +622,76 @@ common unit test frameworks.
...
+duplicating some the code. For example, for each test a new a dead lock check
```
Comment (ianvernon) [inline Documentation/contributing.rst]: A good example of using a global ``JustAfterEach`` function is for deadlock detection, which checks the Cilium logs for deadlocks that may have occurred in the duration of the tests.

File: Documentation/contributing.rst [hunk 3]
```diff
@@ -622,6 +622,76 @@ common unit test frameworks.
...
+runs to validate that nothing wrong happens during the execution of the test,
+that use to be a global ``JustAfterEach`` that validates after each test.
+
+AfterFailed
+^^^^^^^^^^^
+
+This method will run before all ``AfterEach`` and after ``JustAfterEach`` this
```
Comment (ianvernon) [inline Documentation/contributing.rst]: and after ``JustAfterEach``. This function is...

File: Documentation/contributing.rst [hunk 4]
```diff
@@ -622,6 +622,76 @@ common unit test frameworks.
...
+function is only called when the test failed. The main reason of this code is to
```
Comment (ianvernon) [inline Documentation/contributing.rst]: This construct is used to gather logs, the status of Cilium, etc, which provide data for analysis when tests fail.

File: Documentation/contributing.rst [hunk 5]
```diff
@@ -622,6 +622,76 @@ common unit test frameworks.
...
+gather logs, cilium status, etc.. This helps to do some analysis when the system fails.
+
+Global execution
```
Comment (ianvernon) [inline Documentation/contributing.rst]: Global execution is a bit vague. Maybe something like `Example Test Layout`? I'm open to suggestions here.

File: Documentation/contributing.rst [hunk 6]
```diff
@@ -622,6 +622,76 @@ common unit test frameworks.
...
+^^^^^^^^^^^^^^^^
+
+Here is an example of the complete test case flow and when these methods will run:
```
Comment (ianvernon) [inline Documentation/contributing.rst]: Here is an example layout of how a test may be written with the aforementioned constructs:

File: Documentation/contributing.rst [hunk 7]
```diff
@@ -622,6 +622,76 @@ common unit test frameworks.
...
+
+Test description diagram:
+::
+
+	Describe
+	BeforeAll(A)
```
Comment (ianvernon) [inline Documentation/contributing.rst]: Indent things below `Describe` to show they are contained within the `Describe`.

File: Documentation/contributing.rst [hunk 8]
```diff
@@ -622,6 +622,76 @@ common unit test frameworks.
...
+	AfterAll(A)
+	AfterFailed(A)
+	AfterEach(A)
+	JustAfterEach(A)
+	TEST1
```
Comment (ianvernon) [inline Documentation/contributing.rst]: For consistency, call these `TESTA1`, `TESTA2`, etc.

File: Documentation/contributing.rst [hunk 9]
```diff
@@ -653,6 +723,60 @@ saved, in case of a failing test an exhaustive data will be added.
...
+code, and run ginkgo using ``dlv``.
+
+Example how to run ginkgo using ``dlv``:
+
+::
+
+	dlv test . -- --ginkgo.focus="Runtime" -ginkgo.v=true --cilium.provision=false
+
+
+Running test on other vms
```
Comment (ianvernon) [inline Documentation/contributing.rst]: Running Tests in Other VMs

File: Documentation/contributing.rst [hunk 10]
```diff
@@ -653,6 +723,60 @@ saved, in case of a failing test an exhaustive data will be added.
...
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you want to run test in a different VM, you can use ``--cilium.SSHConfig``
```
Comment (ianvernon) [inline Documentation/contributing.rst]: ... use the ...

File: Documentation/contributing.rst [hunk 11]
```diff
@@ -653,6 +723,60 @@ saved, in case of a failing test an exhaustive data will be added.
...
+option to connect over ssh to another server. Cilium source code should be
```
Comment (ianvernon) [inline Documentation/contributing.rst]: to provide the SSH configuration of the endpoint on which tests will be ran.

Comment (ianvernon) [inline Documentation/contributing.rst]: The tests presume the following on the remote instance: 
* Cilium source code is located in the directory ``/home/vagrant/go/src/github.com/cilium/cilium/``.
* Cilium is installed and running.

File: Documentation/contributing.rst [hunk 12]
```diff
@@ -653,6 +723,60 @@ saved, in case of a failing test an exhaustive data will be added.
...
+located on ``/home/vagrant/go/src/github.com/cilium/cilium/`` and test will run
+over the given SSH Connection. Also the target server should have Cilium
+installed and running.
+
+The ssh connection need to be defined as a ``ssh-config`` file and need to have
+the following targets:
+
+- runtime: To run runtime test.
+- k8s{1..2}-${K8S_VERSION}: to run Kubernetes test. It needs a Kubernetes installation
```
Comment (ianvernon) [inline Documentation/contributing.rst]: Kubernetes tests. These instances must have Kubernetes installed and running as a prerequisite for running tests.

File: Documentation/contributing.rst [hunk 13]
```diff
@@ -653,6 +723,60 @@ saved, in case of a failing test an exhaustive data will be added.
...
+If you want to run test in a different VM, you can use ``--cilium.SSHConfig``
+option to connect over ssh to another server. Cilium source code should be
+located on ``/home/vagrant/go/src/github.com/cilium/cilium/`` and test will run
+over the given SSH Connection. Also the target server should have Cilium
+installed and running.
+
+The ssh connection need to be defined as a ``ssh-config`` file and need to have
+the following targets:
+
+- runtime: To run runtime test.
```
Comment (ianvernon) [inline Documentation/contributing.rst]: test --> tests

**Review by ianvernon — CHANGES_REQUESTED** [review]

File: Documentation/contributing.rst [hunk 14]
```diff
@@ -653,6 +724,62 @@ saved, in case of a failing test an exhaustive data will be added.
...
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you want to run test in a different VM, you can use ``--cilium.SSHConfig`` to
+provide the SSH configuration of the endpoint on which tests will be ran. The
+tests presume the following on the remote instance:
+
+- Cilium source code is located in the directory ``/home/vagrant/go/src/github.com/cilium/cilium/``.
+- Cilium is installed and running.
+
+The ssh connection need to be defined as a ``ssh-config`` file and need to have
```
Comment (ianvernon) [inline Documentation/contributing.rst]: need --> needs

File: Documentation/contributing.rst [hunk 15]
```diff
@@ -653,6 +724,62 @@ saved, in case of a failing test an exhaustive data will be added.
...
+the following targets:
+
+- runtime: To run runtime tests
+- k8s{1..2}-${K8S_VERSION}: to run Kubernetes tests. Kubernetes tests. These
```
Comment (ianvernon) [inline Documentation/contributing.rst]: > to run Kubernetes tests. Kubernetes tests.

remove " Kubernetes tests" . Please be more careful with proofreading your changes.

File: Documentation/contributing.rst [hunk 16]
```diff
@@ -622,6 +622,77 @@ common unit test frameworks.
...
+^^^^^^^^^^^^^^^^^^^^^
+
+Here is an example layout of how a test may be written with the aforementioned
+constructs:
+
+Test description diagram:
+::
+
+	Describe
+        BeforeAll(A)
```
Comment (ianvernon) [inline Documentation/contributing.rst]: Indentation doesn't look consistent with things under `Describe` vs. those in `Context` on GitHub.

File: Documentation/contributing.rst [hunk 17]
```diff
@@ -622,6 +622,77 @@ common unit test frameworks.
...
+            TESTB1
+            TESTB2
+            TESTB3
+
+
+Test execution flow:
+::
+
+    Describe
+    BeforeAll
```
Comment (ianvernon) [inline Documentation/contributing.rst]: Indent things under `Describe`.

**Review by ianvernon — CHANGES_REQUESTED** [review]

**Comment by eloycoto** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]

**Pull request merged**